src/ptrace_syscall_info.rs

```diff
@@ -0,0 +1,169 @@
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct RawInfo {
+    op: OpType,
+    arch: u32,
+    instruction_pointer: u64,
+    stack_pointer: u64,
+    data: RawData,
+}
+
+/// See man ptrace (linux) for reference.
+#[derive(Copy, Clone, Debug)]
+pub struct SyscallInfo {
+    pub arch: u32,
+    pub instruction_pointer: u64,
+    pub stack_pointer: u64,
+    pub op: SyscallOp,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum SyscallOp {
+    Entry {
+        nr: u64,
+        args: [u64; 6],
+    },
+    Exit {
+        rval: i64,
+        is_error: u8,
+    },
+    Seccomp {
+        nr: u64,
+        args: [u64; 6],
+        ret_data: u32,
+    },
+    None,
+}
+
+fn parse_raw_data(info: RawInfo) -> Result<SyscallOp> {
+    let op = unsafe {
+        match info.op {
+            OpType::PTRACE_SYSCALL_INFO_NONE => SyscallOp::None,
+            OpType::PTRACE_SYSCALL_INFO_ENTRY => SyscallOp::Entry {
+                nr: info.data.entry.nr,
+                args: info.data.entry.args,
+            },
+            OpType::PTRACE_SYSCALL_INFO_EXIT => SyscallOp::Exit {
+                rval: info.data.exit.rval,
+                is_error: info.data.exit.is_error,
+            },
+            OpType::PTRACE_SYSCALL_INFO_SECCOMP => SyscallOp::Seccomp {
+                nr: info.data.seccomp.nr,
+                args: info.data.seccomp.args,
+                ret_data: info.data.seccomp.ret_data,
+            },
+            _ => bail!("unknown ptrace_syscall_info.op: {:?}", info.op),
+        }
+    };
+
+    Ok(op)
+}
+
+fn parse_raw_info(raw: RawInfo) -> Result<SyscallInfo> {
+    let info = SyscallInfo {
+        arch: raw.arch,
+        instruction_pointer: raw.instruction_pointer,
+        stack_pointer: raw.stack_pointer,
+        op: parse_raw_data(raw)?,
+    };
+    Ok(info)
+}
+
+pub fn get_syscall_info(pid: Pid) -> Result<SyscallInfo> {
+    let mut info = MaybeUninit::<RawInfo>::zeroed();
+    let ret = unsafe {
+        libc::ptrace(
+            PTRACE_GET_SYSCALL_INFO,
+            pid,
+            size_of::<RawInfo>(),
+            info.as_mut_ptr(),
+        )
+    };
+    if ret <= 0 {
+        bail!("ptrace get syscall info error: {}", ret);
+    }
+    let info = unsafe { info.assume_init() };
+    assert!(
+        (info.op == OpType::PTRACE_SYSCALL_INFO_NONE
+            && size_of::<RawInfo>() - size_of::<RawData>() == ret as usize)
+            || (size_of::<RawInfo>() == ret as usize)
+    );
+    let info = try_with!(
+        parse_raw_info(info),
+        "cannot understand ptrace(PTRACE_GET_SYSCALL_INFO) response"
+    );
+    Ok(info)
+}
+
+#[cfg(test)]
+mod test {
+    #[test]
+    fn assert_struct_sizes() {
+        use super::*;
+        //assert_eq!(size_of::<RawInfo>(), 0); // for linux <= v5.2
+
+        assert_eq!(size_of::<RawInfo>(), 88); // for linux <= v5.10
+
+        //assert_eq!(size_of::<RawInfo>(), 84); // for linux >= v5.11
+    }
+
+    #[test]
+    fn check_linux_version() {
```

Comment:
I think we need to make this a runtime check because build machine != target machine.

Reply:
We would need one, if we were to support linux versions other than v5.3-v5.10. But ptrace_syscall_info as of now supports only  linux v5.3-v5.10. In those versions the kernel doesn't answer with valuable information anyways and therefore this code is currently never run. 

i could extract it ptrace_syscall_info to another branch in case we have a use for it later.

Reply:
I mean, does it not work with 5.11 as well, but just show less information than the kernel is capable of?